uring/jvm/src/main/scala/fs2/io/uring/unsafe/UringSystem.scala

```diff
@@ -0,0 +1,248 @@
+
+  override def poll(
+      poller: Poller,
+      nanos: Long,
+      reportFailure: Throwable => Unit
+  ): Boolean =
+    poller.poll(nanos)
+
+  override def needsPoll(poller: Poller): Boolean = poller.needsPoll()
+
+  override def interrupt(targetThread: Thread, targetPoller: Poller): Unit = ()
+
+  private final class ApiImpl(register: (Poller => Unit) => Unit) extends Uring {
+    private[this] val noopRelease: Int => IO[Unit] = _ => IO.unit
+
+    def call(
+        op: Byte,
+        flags: Int,
+        rwFlags: Int,
+        fd: Int,
+        bufferAddress: Long,
+        length: Int,
+        offset: Long
+    ): IO[Int] =
+      exec(op, flags, rwFlags, fd, bufferAddress, length, offset)(noopRelease)
+
+    def bracket(
+        op: Byte,
+        flags: Int,
+        rwFlags: Int,
+        fd: Int,
+        bufferAddress: Long,
+        length: Int,
+        offset: Long
+    )(release: Int => IO[Unit]): Resource[IO, Int] =
+      Resource.makeFull[IO, Int](poll =>
+        poll(exec(op, flags, rwFlags, fd, bufferAddress, length, offset)(release(_)))
+      )(release)
+
+    private def exec(
+        op: Byte,
+        flags: Int,
+        rwFlags: Int,
+        fd: Int,
+        bufferAddress: Long,
+        length: Int,
+        offset: Long
+    )(release: Int => IO[Unit]): IO[Int] = IO.cont {
+      new Cont[IO, Int, Int] {
+        def apply[F[_]](implicit
+            F: MonadCancelThrow[F]
+        ): (Either[Throwable, Int] => Unit, F[Int], IO ~> F) => F[Int] = { (resume, get, lift) =>
+          F.uncancelable { poll =>
+            println("[EXEC]: Entering exec")
+            println("THREAD:" + Thread.currentThread().getName)
+            val submit: IO[Short] = IO.async_[Short] { cb =>
+              register { ring =>
+                val id = ring.getSqe(resume)
+                ring.getSq().enqueueSqe(op, flags, rwFlags, fd, bufferAddress, length, offset, id)
+                cb(Right(id))
+                println("[EXEC]: Leaving exec")
+              }
+            }
+
+            lift(submit)
+              .flatMap { id =>
+                F.onCancel(
+                  poll(get),
+                  lift(cancel(id)).ifM(
+                    F.unit,
+                    // if cannot cancel, fallback to get
+                    get.flatMap { rtn =>
+                      if (rtn < 0) F.raiseError(IOExceptionHelper(-rtn))
+                      else lift(release(rtn))
+                    }
+                  )
+                )
+              }
+              .flatTap(e => F.raiseWhen(e < 0)(IOExceptionHelper(-e)))
+          }
+
+        }
+      }
+    }
+
+    private[this] def cancel(
+        id: Short
+    ): IO[Boolean] =
+      IO.async_[Boolean] { cb =>
+        register { ring =>
+          val wasCancel: Boolean =
+            !ring.getSq().enqueueSqe(OP.IORING_OP_ASYNC_CANCEL, 0, 0, -1, 0, 0, 0, id)
+          cb(Right(wasCancel))
+        }
+      }
+  }
```

Comment:
I don't think this is correct. The `Boolean` that `enqueueSqe()` is actually not what we are interested in—we are interested in the return code from this op, which indicates whether or not it canceled.

We need to treat this like an ordinary operation submission i.e. obtain an id, associate it with the callback, etc. The only difference in this case is that requesting cancelation is not cancelable.